Make toggles available to build errors

orangejenny opened the pull request:
## Product Description
No ticket, fixes issue #16 in [this sheet](https://docs.google.com/spreadsheets/d/1s8xOwCntMk4JiiL5AbdNi2zwhlY-FpteK44DbarYeLg/edit#gid=0), where the build error for using multi-select with "select parent first" should show different error text depending on whether or not the NON_PARENT_MENU_SELECTION flag is on.

## Feature Flag
USH: Allow selecting multiple cases from the case list
Allow selecting of module of any case-type in select-parent workflow

## Safety Assurance

### Safety story
Not requesting QA. I've tested the bug in question locally and gone through the other workflows affected (validating a module, validating a form, making a new build, building a ccz for app preview).

### Automated test coverage

There are tests for build errors, but I don't think there are any for this UI-level code.

### QA Plan

Not requesting QA, see above.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Added toggle utility functions

orangejenny committed:
Updated build_errors rendering to accept full list of toggles

esoergel commented on corehq/toggles/__init__.py:
```diff
@@ -563,6 +563,30 @@ def toggles_enabled_for_user(username):
     }
 
 
+@quickcache(["email"], timeout=24 * 60 * 60, skip_arg=lambda _: settings.UNIT_TESTING)
```
With a timeout like this, we probably want explicit validation, otherwise when you enable a toggle, you'd have to wait 24 hours for it to take effect.

esoergel replied:
Just took a look at the code - didn't realize this was following an existing pattern (I hadn't seen `NAMESPACE_EMAIL_DOMAIN` before).  Those are cleared here:
https://github.com/dimagi/commcare-hq/blob/4c8bbe4093fc3501899480813ead8cb58c2222dd/corehq/apps/toggle_ui/views.py#L286-L294

orangejenny replied:
Good catch, thank you. Fixed in cfe40281c42 and updated the function name to match the namespace constant.

orangejenny committed:
Fixed test

Test test_download_file_bad_xform_404 uses an AnonymousUser, which doesn't have an email attribute.

esoergel reviewed (changes requested):
looks g2m after cache invalidation

orangejenny committed:
Added cache invalidation and updated naming

snopoke reviewed (approved):
Looks good. One comment not blocking.

Also thinking that this could be used by the `toggle_enabled` template tag.

orangejenny committed:
Added email toggles to SessionDetailsView

orangejenny committed:
Updated util functions and callers

orangejenny commented on corehq/toggles/__init__.py:
```diff
@@ -563,6 +563,30 @@ def toggles_enabled_for_user(username):
+    }
+
+
+def toggles_enabled_for_request(request):
```
Made a couple of changes to toggle utils in cf1a7263477. The annoying thing is that some places need toggle names while others need slugs. Now all of the getter functions are based on names and there's a `toggle_names_to_slugs` function to convert when needed.

snopoke commented on corehq/apps/linked_domain/local_accessors.py:
```diff
@@ -46,7 +46,8 @@ def get_enabled_toggles_and_previews(domain):
 def get_enabled_toggles(domain):
-    return list(toggles.toggles_dict(domain=domain))
+    enabled = toggles.toggles_enabled_for_domain(domain)
+    return {slug: True for slug in toggles.toggles_names_to_slugs(enabled)}
```
this was returning a list whereas now it's returning a dict. I think it should be:
```suggestion
    return toggles.toggles_names_to_slugs(enabled)
```

snopoke commented on corehq/apps/hqwebapp/session_details_endpoint/views.py:
```diff
@@ -68,7 +68,7 @@ def post(self, request, *args, **kwargs):
             domains.update(EnterprisePermissions.get_domains(member_domain))
 
-        enabled_toggles = toggles_enabled_for_user(user.username) | toggles_enabled_for_domain(domain)
+        enabled_toggles = toggles_enabled_for_request(request)
```
I don't think you can pass in `request` here since it doesn't have a `domain` or `user` attribute

snopoke reviewed (commented):
I think some bits of the latest changes aren't correct.

For the same of the bug you were fixing perhaps it's better to revert to previous state and leave the 'standardisation' for another PR. Your call.

orangejenny committed:
Revert "Updated util functions and callers"

This reverts commit cf1a7263477b5cac21f879e1f087ad89d884ec1b.

orangejenny committed:
Revert "Added email toggles to SessionDetailsView"

This reverts commit f35ced54e800f90c69684a546a41e1307f03778d.

orangejenny commented:
@snopoke Yeah, this is getting messy. I reverted the last two commits, so this PR should be good now. I'll look at branching a separate PR off of this one.
